bzlmod/canonical-repo-name-injection/BUILD

```diff
@@ -0,0 +1,204 @@
+# Copyright 2024 EngFlow Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+load(
+    "//:repo-names.bzl",
+    "canonical_repo",
+    "workspace_root",
+    "repo_name_variable",
+    "repo_dir_variable",
+    "gen_js_constants",
+)
+load("@bazel_skylib//rules:common_settings.bzl", "string_flag")
+load("@rules_js//js:defs.bzl", "js_binary")
+
+# This value propagates through the rules below.
+# Try it with other targets of your choosing!
+repo_target = "@pnpm"
+
+# The following value inspired the `repo_name_variable` rule.
+# - A macro executes during the loading phase and doesn't see resolved aliases.
+# - A rule executes during the analysis phase, after aliases have resolved.
+#repo_target = "@bison_v3.3.2//bin:bison"
+
+# Demonstrates the macro-based approach.
+genrule(
+    name = "repo-macros",
+    outs = ["repo-macros.txt"],
+    cmd = "printf '%s\n  canonical_name: %s\n  workspace_root: %s\n' >$@" % (
+      repo_target, canonical_repo(repo_target), workspace_root(repo_target)
+    )
+)
+
+# Demonstrates the custom Make variable-based approach. Depends on the
+# ":repo-target" alias to show how the variables have access to the underlying
+# target.
+genrule(
+    name = "repo-vars",
+    outs = ["repo-vars.txt"],
+    cmd = """printf '%%s\n' \
```

Comment:
This might need to be a raw string literal (`r"""printf...`) so the Starlark interpreter doesn't eat the `\` characters. Though if it works already, perhaps no need to change.

Reply:
I went ahead and changed them, so that the resulting `cmd` strings produce what I _intended_, not just something that happened to work.

I checked this by extracting the `cmd` into a `repo_vars_cmd` variable, then calling `fail(repo_vars_cmd)` and running `bazel build //...`. Before this change, the resulting script looked like:

```txt
Error in fail: printf '%s
'           'https://github.com/pnpm'           '  repo-name: $(repo-name)'           '  repo-dir:  $(repo-dir)' >$@
```

And after:

```txt
Error in fail: printf '%s\n' \
          'https://github.com/pnpm' \
          '  repo-name: $(repo-name)' \
          '  repo-dir:  $(repo-dir)' >$@
```